server/src/main/java/org/apache/druid/server/coordinator/helper/SegmentCompactorUtil.java

```diff
@@ -20,21 +20,36 @@
 package org.apache.druid.server.coordinator.helper;
 
 import com.google.common.base.Preconditions;
+import org.apache.druid.timeline.DataSegment;
 import org.joda.time.Interval;
 
+import javax.annotation.Nullable;
+import java.util.List;
+
 /**
  * Util class used by {@link DruidCoordinatorSegmentCompactor} and {@link CompactionSegmentSearchPolicy}.
  */
 class SegmentCompactorUtil
 {
-  static boolean isCompactibleSize(long targetBytes, long currentTotalBytes, long additionalBytes)
-  {
-    return currentTotalBytes + additionalBytes <= targetBytes;
-  }
+  /**
+   * The allowed error rate of the segment size after compaction.
+   * Its value is determined experimentally.
+   */
+  private static final double ALLOWED_ERROR_OF_SEGMENT_SIZE = .2;
 
-  static boolean isCompactibleNum(int numTargetSegments, int numCurrentSegments, int numAdditionalSegments)
+  static boolean needsCompaction(@Nullable Long targetCompactionSizeBytes, List<DataSegment> candidates)
   {
-    return numCurrentSegments + numAdditionalSegments <= numTargetSegments;
+    if (targetCompactionSizeBytes == null) {
+      // If targetCompactionSizeBytes is null, we have no way to check that the given segments need compaction or not.
+      return true;
+    }
+    final double minAllowedSize = targetCompactionSizeBytes * (1 - ALLOWED_ERROR_OF_SEGMENT_SIZE);
+    final double maxAllowedSize = targetCompactionSizeBytes * (1 + ALLOWED_ERROR_OF_SEGMENT_SIZE);
+
+    return candidates
+        .stream()
+        .filter(segment -> segment.getSize() < minAllowedSize || segment.getSize() > maxAllowedSize)
+        .count() > 1;
```

Comment:
nit: I think this should use `anyMatch` instead of `filter` + `count` since the former might not have to materialize the entire set upon reaching first match

Reply:
ignore ^